### ISSUE TITLE
[Q] practical tips for batchnorm?

### ISSUE THREAD

**Issue by ywpkwon** [tensorpack/tensorpack#1343]
Hello from a big fan of tensorpack. 

I have a practical question about BatchNorm. As far as I know, special care should be needed when using BatchNorm with multiple GPUs (is it right?), and tensorpack can handle natively.

I'm eager to know some possible mistakes. I would really appreciate it if you could provide some practical advice.

Sometimes I borrow a backbone defined using `tf.slim` or `tf.layers`.
For example, 
```python
class Model(ModelDesc):
    ...
    def build_graph(self, ...):
       batch_norm_params = {
            'decay': 0.9997,          # decay for the moving averages.
            'epsilon': 0.001,          # epsilon to prevent 0s in variance.
            'updates_collections': tf.GraphKeys.UPDATE_OPS,   # collection containing update_ops.
        }
        with arg_scope([layers.conv2d, layers_lib.fully_connected],
                       weights_regularizer=regularizers.l2_regularizer(0.00004),
                       activation_fn=tf.nn.relu,
                       normalizer_fn=layers_lib.batch_norm,
                       normalizer_params=batch_norm_params): 
               some_backbone_here()
        ...
        return loss
```
Except the definition, I'm training with `AutoResumeTrainConfig`, `launch_train_with_config`.

In this case, can I expect correct BatchNorm behavior with multiple GPU?
Or, should I replace every layer (`tf.layers.conv2d`, `tf.layers.batchnorm`, etc) with Tensorpack's layers (`Conv2D`, `BatchNorm`)?

**Comment by ppwwyyxx** [comment]
> can I expect correct BatchNorm behavior with multiple GPU?

Depends on what behavior you want exactly..

The way you use `tf.slim` will work the same as using tensorpack's `BatchNorm`.
